docker-installer/install_cni_plugins.sh

```diff
@@ -0,0 +1,59 @@
+#!/bin/sh
+
+set -e
+
+if [ $# -ne 2 ]; then
+  echo "USAGE: $0 source-dir dest-dir"
+  exit 1
+fi
+
+SRC=$1
+DST=$2
+
+install_cni_plugin() {
+	NAME=$1
+	MD5=$( md5sum $SRC/$NAME | awk '{ print $1 }' )
+	if [ -e $DST/$NAME ]; then
+		if [ ! -e $DST/$NAME.by_cni_installer_image ]; then
+			# The file already exists but there's no marker that
+			# it was installed by this installer -> keep untouched
+			echo "* '$NAME' ignored (exists but not installed by me)"
+			return
+		fi
+		OTHER_MD5=$( md5sum $DST/$NAME | awk '{ print $1 }' )
+		INSTALLED_MD5=$( cat $DST/$NAME.by_cni_installer_image )
+
+		if [ "$OTHER_MD5" != "$INSTALLED_MD5" ]; then
+			# The file was previously installed by this installer
+			# but later changed -> keep untouched
+			echo "* '$NAME' ignored (previously installed by me but changed by someone else)"
+			return
+		fi
+
+		if [ "$OTHER_MD5" == "$MD5" ]; then
+			# The file was previously installed by this installer
+			# and is up-to-date
+			echo "* '$NAME' is up-to-date"
+			return
+		fi
+
+		# The file was previously installed by this installer
+		# but needs an update
+		cp -a $SRC/$NAME $DST/$NAME
```

Comment:
We should make these writes atomic, so can you copy to a temporary file (in the same directory) then `mv` them? That way there's no chance of reading a partially-written executable.

I actually had to fix this bug in a different container I maintain, so it can definitely happen :-/

Reply:
Another option would be to name them $NAME.by_cni_installer (rather than creating that marker file) and set a symlink. That makes handling of installation/updates easier:
* if $NAME not exists -> install it
* if $NAME exists and is the correct symlink -> probably update the binary
* if $NAME exists and is not a symlink (or has wrong target) -> not our file, keep it

Reply:
You still have the same problem, though - if you write $NAME.by_cni_installer in place, you could still get a partial read.